src/cdomains/valueDomain.ml

```diff
@@ -673,20 +765,45 @@ struct
 
 
   (* Funny, this does not compile without the final type annotation! *)
-  let rec eval_offset (ask: Q.ask) f (x: t) (offs:offs) (exp:exp option) (v:lval option): t =
-    let rec do_eval_offset (ask:Q.ask) f (x:t) (offs:offs) (exp:exp option) (l:lval option) (o:offset option) (v:lval option):t =
+  let rec eval_offset (ask: Q.ask) f (x: t) (offs:offs) (exp:exp option) (v:lval option) (t:typ): t =
+    let rec do_eval_offset (ask:Q.ask) f (x:t) (offs:offs) (exp:exp option) (l:lval option) (o:offset option) (v:lval option) (t:typ): t =
       match x, offs with
-      | `Blob c, `Index (_, ox) ->
+      | `Blob((va, _, orig) as c), `Index (_, ox) ->
         begin
           let l', o' = shift_one_over l o in
-          do_eval_offset ask f (Blobs.value c) ox exp l' o' v
+          let ev = do_eval_offset ask f (Blobs.value c) ox exp l' o' v t in
+          if orig then (* This Blob came from malloc *) 
+            ev 
+          else (* This Blob came from calloc *)
+            if ev = `Bot then
+              zero_init_value t (* This should be zero initialized *)
+            else
+              ev (* This already contains some value *)
```

Comment:
This logic is copy-pasted three times here. Might be better to have a helper function for it.